File: labs/daemon/modules/relay/weighted_router.go
package relay

import (
	"math"
	"sync"
)

type EgressRoute struct {
	ID            string
	Endpoint      string
	Weight        int
	CurrentWeight int
	Healthy       bool
}

type WeightedEgressRouter struct {
	mu     sync.Mutex
	routes []*EgressRoute
}

func NewWeightedEgressRouter() *WeightedEgressRouter {
	return &WeightedEgressRouter{
		routes: make([]*EgressRoute, 0),
	}
}

func (r *WeightedEgressRouter) AddRoute(id, endpoint string, weight int) {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.routes = append(r.routes, &EgressRoute{
		ID:            id,
		Endpoint:      endpoint,
		Weight:        weight,
		CurrentWeight: 0,
		Healthy:       true,
	})
}

func (r *WeightedEgressRouter) SetHealth(id string, healthy bool) {
	r.mu.Lock()
	defer r.mu.Unlock()

	for _, rt := range r.routes {
		if rt.ID == id {
			rt.Healthy = healthy
			if !healthy {
				rt.CurrentWeight = 0
			}
			break
		}
	}
}

func (r *WeightedEgressRouter) NextRoute() (string, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()

	totalHealthyWeight := 0
	for _, rt := range r.routes {
		if rt.Healthy && rt.Weight > 0 {
			totalHealthyWeight += rt.Weight
		}
	}

	if totalHealthyWeight <= 0 {
		return "", false
	}

	var bestRoute *EgressRoute
	maxWeight := math.MinInt32

	for _, rt := range r.routes {
		if !rt.Healthy || rt.Weight <= 0 {
			continue
		}
		rt.CurrentWeight += rt.Weight
		if rt.CurrentWeight > maxWeight {
			maxWeight = rt.CurrentWeight
			bestRoute = rt
		}
	}

	if bestRoute != nil {
		bestRoute.CurrentWeight -= totalHealthyWeight
		return bestRoute.Endpoint, true
	}

	return "", false
}

File: labs/control-ui/api-shadow/weighted_router.ts
/**
 * Smooth Weighted Round-Robin (SWRR) Router
 */
export interface EgressRoute {
  id: string;
  endpoint: string;
  weight: number;
  currentWeight: number;
  healthy: boolean;
}

export class WeightedEgressRouter {
  private routes: EgressRoute[] = [];

  addRoute(id: string, endpoint: string, weight: number): void {
    this.routes.push({ id, endpoint, weight, currentWeight: 0, healthy: true });
  }

  setHealth(id: string, healthy: boolean): void {
    const r = this.routes.find((rt) => rt.id === id);
    if (r) {
      r.healthy = healthy;
      if (!healthy) r.currentWeight = 0;
    }
  }

  nextRoute(): string | null {
    const totalHealthy = this.routes
      .filter((r) => r.healthy && r.weight > 0)
      .reduce((acc, r) => acc + r.weight, 0);

    if (totalHealthy <= 0) return null;

    let best: EgressRoute | null = null;
    let maxWeight = -Infinity;

    for (const r of this.routes) {
      if (!r.healthy || r.weight <= 0) continue;
      r.currentWeight += r.weight;
      if (r.currentWeight > maxWeight) {
        maxWeight = r.currentWeight;
        best = r;
      }
    }

    if (best) {
      best.currentWeight -= totalHealthy;
      return best.endpoint;
    }
    return null;
  }
}
